vm/interpreter/src/lib.rs

```diff
@@ -1,130 +1,319 @@
 // Copyright 2020 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
+use actor::{ACCOUNT_ACTOR_CODE_ID, REWARD_ACTOR_ADDR};
 use address::Address;
 use blocks::Tipset;
 use clock::ChainEpoch;
 use default_runtime::{internal_send, DefaultRuntime};
+use forest_encoding::Cbor;
 use ipld_blockstore::BlockStore;
+use log::warn;
 use message::{Message, MessageReceipt, SignedMessage, UnsignedMessage};
 use num_bigint::BigUint;
-use num_traits::cast::ToPrimitive;
-use runtime::DefaultSyscalls;
-use vm::ActorError;
-use vm::{ExitCode, Serialized, StateTree};
-
+use num_traits::Zero;
+use runtime::Syscalls;
+use vm::{price_list_by_epoch, ActorError, ExitCode, Serialized, StateTree};
 /// Interpreter which handles execution of state transitioning messages and returns receipts
 /// from the vm execution.
-pub struct VM<'a, ST: StateTree, DB: BlockStore> {
+pub struct VM<'a, ST, DB, SYS>
+where
+    ST: StateTree,
+    DB: BlockStore,
+    SYS: Syscalls + Copy,
+{
     state: ST,
     store: &'a DB,
     epoch: ChainEpoch,
+    syscalls: SYS,
     // TODO: missing fields
 }
 
-impl<'a, ST: StateTree, DB: BlockStore> VM<'a, ST, DB> {
-    pub fn new(state: ST, store: &'a DB, epoch: ChainEpoch) -> Self {
+impl<'a, ST, DB, SYS> VM<'a, ST, DB, SYS>
+where
+    ST: StateTree,
+    DB: BlockStore,
+    SYS: Syscalls + Copy,
+{
+    pub fn new(state: ST, store: &'a DB, epoch: ChainEpoch, syscalls: SYS) -> Self {
         VM {
             state,
             store,
             epoch,
+            syscalls,
         }
     }
 
+    /// Returns ChainEpoch
+    fn epoch(&self) -> &ChainEpoch {
+        &self.epoch
+    }
```

Comment:
```suggestion
    /// Returns ChainEpoch
    fn epoch(&self) -> ChainEpoch {
        self.epoch
    }
```
ChainEpoch is just a u64, any reason specifically for returning a reference?

Reply:
Not just a habit. I will remove.